packaging/format-docs.py

```diff
@@ -0,0 +1,82 @@
+#!/usr/bin/env python3
+# Copyright 2007-2022 The OpenRA Developers (see AUTHORS)
+# This file is part of OpenRA, which is free software. It is made
+# available to you under the terms of the GNU General Public License
+# as published by the Free Software Foundation, either version 3 of
+# the License, or (at your option) any later version. For more
+# information, see COPYING.
+
+import io
+import sys
+import json
+from collections import OrderedDict
+
+def format_type_name(typeName):
+    name = typeName
+    if name.endswith("Info"):
+        name = name[0:-4]
+
+    return f'[`{name}`](#{name.lower()})'
+
+def format_docs(version, collectionName, types):
+    typesByNamespace = OrderedDict()
+    for currentType in types:
+        if currentType["Namespace"] in typesByNamespace:
+            typesByNamespace[currentType["Namespace"]].append(currentType)
+        else:
+            typesByNamespace[currentType["Namespace"]] = [currentType]
+
+    explanation = ""
+    if collectionName == "TraitInfos":
+        explanation = "all traits with their properties and their default values plus developer commentary"
+    elif collectionName == "WeaponTypes":
+        explanation = "a template for weapon definitions as well as its contained types (warheads and projectiles) with default values and developer commentary"
+    elif collectionName == "SequenceTypes":
+        explanation = "all sequence types with their properties and their default values plus developer commentary"
+
+    print(f"This documentation is aimed at modders and has been automatically generated for version `{version}` of OpenRA. " +
+				"Please do not edit it directly, but instead add new `[Desc(\"String\")]` tags to the source code.\n")
+
+    print(f"Listed below are {explanation}:\n")
+
+    for namespace in typesByNamespace:
+        print(f'## {namespace}\n')
+
+        for currentType in typesByNamespace[namespace]:
+            print(f'### {currentType["Name"]}\n')
+
+            if currentType["Description"]:
+                print(f'#### {currentType["Description"]}\n')
+
+            if "InheritedTypes" in currentType and currentType["InheritedTypes"]:
+                print("Inherits from: " + ", ".join([format_type_name(x) for x in currentType["InheritedTypes"]]) + '.\n')
+
+            if "RequiresTraits" in currentType and currentType["RequiresTraits"]:
+                print("Requires trait(s): " + ", ".join([format_type_name(x) for x in currentType["RequiresTraits"]]) + '.\n')
+
+            if len(currentType["Properties"]) > 0:
+                hasAttributes = "OtherAttributes" in currentType["Properties"][0]
```

Comment:
This is always true. See `TiberianSunRefinery`, it doesn't have any attributes on any of its values. 

This also fails at a lower part of the code

Reply:
This is actually here to allow us to make a distinction between Trait docs, Weapon docs and Sequence docs, because the first two include attributes while the third one doesn't and we want to reuse the "table generation" code. I really should add a comment explaining that though because it did take me a while to figure out what you mean.

Reply:
ah ok, then a comment is definitely needed